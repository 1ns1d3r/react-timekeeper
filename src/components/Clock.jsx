import React, { PropTypes } from 'react'
import { spring, TransitionMotion, presets } from 'react-motion';

// TODO - remove this depdency
import css, {loop} from 'reactcss'

import calcRelativePositioning from '../helpers/relative'
/*const Clock = ({}) => {
	return <div>Clock: 0 5 10 15 20 25...</div>
}*/

// radius of clock, in px
const CLOCK_RADIUS = 120
const CLOCK_SIZE = CLOCK_RADIUS * 2

// clock hand length, in px
const CLOCK_HAND_LENGTH = 90

// number of actual numbers to display
const NUMBER_INCREMENTS = 12
const NUMBER_INCREMENTS_VALUE = 360 / NUMBER_INCREMENTS

// size of circle surrounding number
const NUMBER_SIZE = 36

// positioning of numbers within circle
const NUMBER_INNER_POSITION = 24

function animationPosition(unit){
	return unit === 'hour' ? NUMBER_INNER_POSITION - 34 : NUMBER_INNER_POSITION + 28;
}

const { cos, sin, atan2 } = Math
const pi = Math.PI

function rad(deg){
	return deg / (180 / pi)
}
function deg(rad){
	return rad * (180 / pi)
}

const CLOCK_DATA = {
	hour: {
		numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
		increments: 12
	},
	minute: {
		// TODO - change this to string
		numbers: ['05', 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, '00'],
		increments: 60
	}
}

/*
	TODO
	- add touchMove
*/

class Clock extends React.Component {
	constructor(props){
		super(props)

		this.state = {}
	}

	render(){
		const props = this.props
		const styles = css({
			default: {
				clockWrapper: {
					// background: 'purple'
					textAlign: 'center',
					padding: '20px 0'
				},
				clock: {
					display: 'inline-block',
					borderRadius: '200px',
					background: 'white',
					width: `${CLOCK_SIZE}px`,
					height: `${CLOCK_SIZE}px`,
					position: 'relative',
					cursor: 'pointer',
				},
				numberPositioning: {
					display: 'inline-block',
					position: 'absolute',
					color: '#898989',
					fontSize: '18px',
					pointerEvents: 'none',

					// background: '#E6F7FF',
					// border: '1px solid green',
					// background: '#00acff',
					borderRadius: '99px',
					width: NUMBER_SIZE,
					height: NUMBER_SIZE,

					textAlign: 'center',
					lineHeight: NUMBER_SIZE + 'px',
					
					zIndex: 5,
				},
				clockHand: {
					position: 'relative'
				}
			},
		});

		function renderNumbersAndClockhand(){
			const unit = props.unit
			const animationItems = [unit === 'hour' ? 'hour' : 'minute'];

			const animationOptions = {
				willEnter(transition){
					return {
						opacity: 0,
						handOpacity: 0,
						translate: animationPosition(transition.data),
					}
				},
				willLeave(transition){
					return {
						opacity: spring(0),
						handOpacity: spring(0),
						translate: spring( animationPosition(transition.data) ),
					}
				},
				styles: animationItems.map(unit => {
					return {
						key: unit,
						style: {
							opacity: spring(1),
							handOpacity: spring(1, {stiffness: 120, damping: 40}),
							translate: spring(NUMBER_INNER_POSITION),
						},
						data: unit
					}
				})
			}

			const handRotation = props[unit] * (360 / CLOCK_DATA[unit].increments)

			return <TransitionMotion {...animationOptions}>
				{interpolatedStyles =>
					<div>
						{interpolatedStyles.map(config => {
							const data = CLOCK_DATA[config.data]
							
							return <div style={{position: 'absolute'}} key={config.data} ref={el => this.clock = el}>
								{data.numbers.map((numberString, i) => {
									const num = i + 1;
									return (
										<span
											key={numberString}
											style={{
												...styles.numberPositioning,
												opacity: config.style.opacity,
												left: sin( rad(num * -NUMBER_INCREMENTS_VALUE - 180) ) * (CLOCK_RADIUS - config.style.translate) + CLOCK_RADIUS - NUMBER_SIZE / 2,
												top: cos( rad(num * -NUMBER_INCREMENTS_VALUE - 180) ) * (CLOCK_RADIUS - config.style.translate) + CLOCK_RADIUS - NUMBER_SIZE / 2,
											}}
										>
											{numberString}
										</span>
									)
								})}
								<svg width={CLOCK_SIZE} height={CLOCK_SIZE} viewBox={`0 0 ${CLOCK_SIZE} ${CLOCK_SIZE}`} xmlns="http://www.w3.org/2000/svg"
									style={{
										...styles.clockHand,
										opacity: config.style.handOpacity,
									}}
								>
									<g transform={`rotate(${handRotation} ${CLOCK_RADIUS} ${CLOCK_RADIUS})`}>
										<line x1={CLOCK_RADIUS} y1={CLOCK_RADIUS} x2={CLOCK_RADIUS} y2={CLOCK_RADIUS - CLOCK_HAND_LENGTH}
											strokeWidth="1"
											stroke="#BCEAFF"
										/>
										<circle cx={CLOCK_RADIUS} cy={CLOCK_RADIUS} r={1.5}
											fill='#BCEAFF'
										/>
										<circle cx={CLOCK_RADIUS} cy={24} r={NUMBER_SIZE / 2}
											fill='#E6F7FF'
										/>
									</g>
								</svg>

							</div>
						})}
					</div>
				}
			</TransitionMotion>
		}


		return (
			<div style={styles.clockWrapper}>
				{/* formatted numbers + clock hand */}
				<div style={styles.clock} onMouseDown={this.mousedown.bind(this)}>
					{ renderNumbersAndClockhand.call(this) }
				</div>
			</div>
		)
	}
	
	handlePoint(clientX, clientY, el){
		// const { minX, minY } = this.state
		// TODO - cache?
		// TODO - fix if parent has absolute positioning
		const { minX, minY } = calcRelativePositioning(this.clock)

		const x = clientX + minX - CLOCK_RADIUS
		const y = -clientY - minY + CLOCK_RADIUS

		const a = atan2(y, x)
		let d = 90 - deg(a)
		if ( d < 0 ){
			d = 360 + d
		}

		// TODO - move this const to state
		const unit = this.props.unit

		const selected = Math.round( d / 360 * CLOCK_DATA[unit].increments )

		if (unit === 'hour'){
			this.props.changeHour(selected)
		} else if (unit === 'minute'){
			this.props.changeMinute(selected)
		}
	}

	mousedown(e){
		this.handlePoint(e.clientX, e.clientY, e.target)

		// bind handlers
		this.dragHandler = this.drag.bind(this)
		this.stopDragHandler = this.stopDragHandler.bind(this)

		// add listeners
		document.addEventListener('mousemove', this.dragHandler, false)
		document.addEventListener('mouseup', this.stopDragHandler, false)
		this.clock.addEventListener('mouseleave', this.stopDragHandler, false)

		// TODO - add touch
	}
	drag(e){
		this.handlePoint(e.clientX, e.clientY)
		e.preventDefault()
		return false
	}

	stopDragHandler(){
		document.removeEventListener('mousemove', this.dragHandler, false)
		document.removeEventListener('mouseup', this.stopDragHandler, false)
		this.clock.removeEventListener('mouseleave', this.stopDragHandler, false)
	}

	componentWillUnmount(){
		// clear any event listeners
		this.stopDragHandler()
	}
}

Clock.propTypes = {
	hour: PropTypes.number.isRequired,
	minute: PropTypes.number.isRequired,
	unit: PropTypes.string.isRequired,
	
	changeUnit: PropTypes.func.isRequired,
	changeHour: PropTypes.func.isRequired,
	changeMinute: PropTypes.func.isRequired
}

export default Clock