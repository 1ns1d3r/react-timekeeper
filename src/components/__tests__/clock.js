import React from 'react'
import renderer from 'react-test-renderer'

import { Clock } from '../Clock';

describe('Clock component', () => {
	const hour = 4;
	const minute = 42;

	const changeHour = jest.fn();
	const changeMinute = jest.fn();

	describe('renders correctly', () => {
		test('when unit is hour', () => {
			const unit = 'hour';
			const tree = renderer.create(
				<Clock
					config={{}}
					unit={unit}
					hour={hour}
					minute={minute}
					changeHour={changeHour}
					changeMinute={changeMinute}
				/>
			).toJSON()
			expect(tree).toMatchSnapshot()
		})
		test('when unit is minute', () => {
			const unit = 'minute';
			const tree = renderer.create(
				<Clock
					config={{}}
					unit={unit}
					hour={hour}
					minute={minute}
					changeHour={changeHour}
					changeMinute={changeMinute}
				/>
			).toJSON()
			expect(tree).toMatchSnapshot()
		})
	})
})