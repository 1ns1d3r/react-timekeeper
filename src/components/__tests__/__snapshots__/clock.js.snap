exports[`Clock component renders correctly when unit is hour 1`] = `
<div
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "background": undefined,
      "borderRadius": "200px",
      "cursor": "pointer",
      "display": "inline-block",
      "height": "240px",
      "position": "relative",
      "width": "240px",
    }
  }>
  <div>
    <div
      style={
        Object {
          "position": "absolute",
        }
      }>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 149.99999999999997,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 18.861561236693873,
            "width": 36,
            "zIndex": 5,
          }
        }>
        1
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 185.1384387633061,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 53.99999999999996,
            "width": 36,
            "zIndex": 5,
          }
        }>
        2
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 198,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 101.99999999999999,
            "width": 36,
            "zIndex": 5,
          }
        }>
        3
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 185.1384387633061,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 150,
            "width": 36,
            "zIndex": 5,
          }
        }>
        4
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 150.00000000000006,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 185.13843876330608,
            "width": 36,
            "zIndex": 5,
          }
        }>
        5
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 102.00000000000003,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 198,
            "width": 36,
            "zIndex": 5,
          }
        }>
        6
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 54,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 185.1384387633061,
            "width": 36,
            "zIndex": 5,
          }
        }>
        7
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 18.861561236693916,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 150.00000000000006,
            "width": 36,
            "zIndex": 5,
          }
        }>
        8
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 6,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 102.00000000000003,
            "width": 36,
            "zIndex": 5,
          }
        }>
        9
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 18.86156123669386,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 54.00000000000007,
            "width": 36,
            "zIndex": 5,
          }
        }>
        10
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 54.00000000000003,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 18.861561236693873,
            "width": 36,
            "zIndex": 5,
          }
        }>
        11
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 101.99999999999997,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 6,
            "width": 36,
            "zIndex": 5,
          }
        }>
        12
      </span>
      <svg
        height={240}
        style={
          Object {
            "opacity": 1,
            "position": "relative",
          }
        }
        viewBox="0 0 240 240"
        width={240}
        xmlns="http://www.w3.org/2000/svg">
        <g
          transform="rotate(120 120 120)">
          <line
            stroke={undefined}
            strokeWidth="1"
            x1={120}
            x2={120}
            y1={120}
            y2={30} />
          <circle
            cx={120}
            cy={120}
            fill={undefined}
            r={1.5} />
          <circle
            cx={120}
            cy={24}
            fill={undefined}
            r={18} />
        </g>
      </svg>
    </div>
  </div>
</div>
`;

exports[`Clock component renders correctly when unit is minute 1`] = `
<div
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "background": undefined,
      "borderRadius": "200px",
      "cursor": "pointer",
      "display": "inline-block",
      "height": "240px",
      "position": "relative",
      "width": "240px",
    }
  }>
  <div>
    <div
      style={
        Object {
          "position": "absolute",
        }
      }>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 149.99999999999997,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 18.861561236693873,
            "width": 36,
            "zIndex": 5,
          }
        }>
        05
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 185.1384387633061,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 53.99999999999996,
            "width": 36,
            "zIndex": 5,
          }
        }>
        10
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 198,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 101.99999999999999,
            "width": 36,
            "zIndex": 5,
          }
        }>
        15
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 185.1384387633061,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 150,
            "width": 36,
            "zIndex": 5,
          }
        }>
        20
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 150.00000000000006,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 185.13843876330608,
            "width": 36,
            "zIndex": 5,
          }
        }>
        25
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 102.00000000000003,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 198,
            "width": 36,
            "zIndex": 5,
          }
        }>
        30
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 54,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 185.1384387633061,
            "width": 36,
            "zIndex": 5,
          }
        }>
        35
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 18.861561236693916,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 150.00000000000006,
            "width": 36,
            "zIndex": 5,
          }
        }>
        40
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 6,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 102.00000000000003,
            "width": 36,
            "zIndex": 5,
          }
        }>
        45
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 18.86156123669386,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 54.00000000000007,
            "width": 36,
            "zIndex": 5,
          }
        }>
        50
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 54.00000000000003,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 18.861561236693873,
            "width": 36,
            "zIndex": 5,
          }
        }>
        55
      </span>
      <span
        style={
          Object {
            "borderRadius": "99px",
            "color": undefined,
            "display": "inline-block",
            "fontSize": "18px",
            "height": 36,
            "left": 101.99999999999997,
            "lineHeight": "36px",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "absolute",
            "textAlign": "center",
            "top": 6,
            "width": 36,
            "zIndex": 5,
          }
        }>
        00
      </span>
      <svg
        height={240}
        style={
          Object {
            "opacity": 1,
            "position": "relative",
          }
        }
        viewBox="0 0 240 240"
        width={240}
        xmlns="http://www.w3.org/2000/svg">
        <g
          transform="rotate(252 120 120)">
          <line
            stroke={undefined}
            strokeWidth="1"
            x1={120}
            x2={120}
            y1={120}
            y2={30} />
          <circle
            cx={120}
            cy={120}
            fill={undefined}
            r={1.5} />
          <circle
            cx={120}
            cy={24}
            fill={undefined}
            r={18} />
          <circle
            cx={120}
            cy={24}
            fill={undefined}
            r={4} />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
